pkg/filesystem/filesystem.go

```diff
@@ -0,0 +1,259 @@
+	return nil
+}
+
+func (c *FileSystem) mountImage(cluster *v2.Cluster) error {
+	var (
+		mountdir = common.DefaultMountCloudImageDir(cluster.Name)
+		upperDir = filepath.Join(mountdir, "upper")
+		driver   = mount.NewMountDriver()
+		err      error
+	)
+	if isMount, _ := mount.GetMountDetails(mountdir); isMount {
+		err = driver.Unmount(mountdir)
+		if err != nil {
+			return fmt.Errorf("%s already mount, and failed to umount %v", mountdir, err)
+		}
+	}
+	if utils.IsFileExist(mountdir) {
+		err = os.RemoveAll(mountdir)
+		if err != nil {
+			return fmt.Errorf("failed to clean %s, %v", mountdir, err)
+		}
+	}
+	//get layers
+	Image, err := c.imageStore.GetByName(cluster.Spec.Image)
+	if err != nil {
+		return err
+	}
+	layers, err := image.GetImageLayerDirs(Image)
+	if err != nil {
+		return fmt.Errorf("get layers failed: %v", err)
+	}
+
+	if err = os.MkdirAll(upperDir, 0744); err != nil {
+		return fmt.Errorf("create upperdir failed, %s", err)
+	}
+	if err = driver.Mount(mountdir, upperDir, layers...); err != nil {
+		return fmt.Errorf("mount files failed %v", err)
+	}
+	return nil
+}
+
+func (c *FileSystem) MountImage(cluster *v2.Cluster) error {
+	return c.mountImage(cluster)
+}
+
+func (c *FileSystem) UnMountImage(cluster *v2.Cluster) error {
+	return c.umountImage(cluster)
+}
+
+func (c *FileSystem) MountRootfs(cluster *v2.Cluster, hosts []string, initFlag bool) error {
+	clusterRootfsDir := common.DefaultTheClusterRootfsDir(cluster.Name)
+	//scp roofs to all Masters and Nodes,then do init.sh
+	if err := mountRootfs(hosts, clusterRootfsDir, cluster, initFlag); err != nil {
+		return fmt.Errorf("mount rootfs failed %v", err)
+	}
+	return nil
+}
+
+func (c *FileSystem) UnMountRootfs(cluster *v2.Cluster) error {
+	//do clean.sh,then remove all Masters and Nodes roofs
+	IPList := append(runtime.GetMasterIPList(cluster), runtime.GetNodeIPList(cluster)...)
+	config := runtime.GetRegistryConfig(common.DefaultTheClusterRootfsDir(cluster.Name), runtime.GetMaster0Ip(cluster))
+	if utils.NotIn(config.IP, IPList) {
+		IPList = append(IPList, config.IP)
+	}
+	if err := unmountRootfs(IPList, cluster); err != nil {
+		return err
+	}
+	return nil
+}
+
+func mountRootfs(ipList []string, target string, cluster *v2.Cluster, initFlag bool) error {
+	config := runtime.GetRegistryConfig(
+		common.DefaultTheClusterRootfsDir(cluster.Name),
+		runtime.GetMaster0Ip(cluster))
+	/*	if err := ssh.WaitSSHToReady(*cluster, 6, ipList...); err != nil {
+		return errors.Wrap(err, "check for node ssh service time out")
+	}*/
+	var wg sync.WaitGroup
+	var flag bool
+	var mutex sync.Mutex
+	src := common.DefaultMountCloudImageDir(cluster.Name)
+	// TODO scp sdk has change file mod bug
+	initCmd := fmt.Sprintf(RemoteChmod, target)
+	for _, ip := range ipList {
+		wg.Add(1)
+		go func(ip string) {
+			defer wg.Done()
+			ssh, err := ssh.GetHostSSHClient(ip, cluster)
+			if err != nil {
+				logger.Error("get host ssh client failed %v", err)
+				mutex.Lock()
+				flag = true
```

Comment:
shall we just return when flag is true? I found that the code continue running in the goroutine. Is it reasonable?

Reply:
Using channel to handle goroutine error